### PR TITLE
Handle 'Cancelled' build status in the web console

### LINKED ISSUE #2144: canceled builds is still showing running icon
osc v0.5-117-gd6036ef
kubernetes v0.16.2-338-gc07896e

running `osc cancel-build` cancels the build and status label gets updated but the icon is still the animated running icon.

build-cancel-icon

also on the overview page, the canceled build looks like a running build

canceled-builds-overview

/cc @spadgett

### PR DESCRIPTION
**Pull request by spadgett**
Fixes #2144

Browse -> Builds

openshift_management_console

Overview

openshift_management_console

### CODE REVIEW

**Comment by spadgett** [comment]
@liggitt please review

File: assets/app/views/_triggers.html [hunk 1]
```diff
@@ -3,9 +3,16 @@
     <span ng-switch="build.status" class="hide-ng-leave">
       <span ng-switch-when="Complete" class="fa fa-check text-success" aria-hidden="true" style="margin-right: 5px;"></span>
       <span ng-switch-when="Failed" class="fa fa-times text-danger" aria-hidden="true" style="margin-right: 5px;"></span>
+      <span ng-switch-when="Cancelled" class="fa fa-ban text-warning" aria-hidden="true"></span>
```
Comment (liggitt) [inline assets/app/views/_triggers.html]: do we know all the states? Seems like we might want a hidden icon to be the default, and only show spinning for "Pending", "Running", etc

File: assets/app/views/_triggers.html [hunk 2]
```diff
@@ -3,9 +3,16 @@
...
       <span ng-switch-default class="fa fa-refresh fa-spin" aria-hidden="true" style="margin-right: 5px;"></span>
     </span>
-    <span>A build of {{build.metadata.labels.buildconfig}} is {{build.status | lowercase}}.</span>
-    <span ng-if="build.status != 'Complete' && build.status != 'Failed' && trigger.imageChangeParams.automatic">  A new deployment will be created automatically once the build completes.</span>
+    <span>
+      A build of {{build.metadata.labels.buildconfig}}
```
Comment (liggitt) [inline assets/app/views/_triggers.html]: probably need another switch
Failed: A build of ... failed
Cancelled: A build of ... was cancelled
Default: A build of ... is lower(status)

File: assets/app/views/_triggers.html (hunk 1 above)
Comment (spadgett) [inline assets/app/views/_triggers.html]: @liggitt Looks like these are the states. We should handle Error as well.

```
// Valid values for BuildPhase.
const (
	// BuildPhaseNew is automatically assigned to a newly created build.
	BuildPhaseNew BuildPhase = "New"

	// BuildPhasePending indicates that a pod name has been assigned and a build is
	// about to start running.
	BuildPhasePending BuildPhase = "Pending"

	// BuildPhaseRunning indicates that a pod has been created and a build is running.
	BuildPhaseRunning BuildPhase = "Running"

	// BuildPhaseComplete indicates that a build has been successful.
	BuildPhaseComplete BuildPhase = "Complete"

	// BuildPhaseFailed indicates that a build has executed and failed.
	BuildPhaseFailed BuildPhase = "Failed"

	// BuildPhaseError indicates that an error prevented the build from executing.
	BuildPhaseError BuildPhase = "Error"

	// BuildPhaseCancelled indicates that a running/pending build was stopped from executing.
	BuildPhaseCancelled BuildPhase = "Cancelled"
)
```

Comment (spadgett) [inline assets/app/views/_triggers.html]: We might also add tooltip help for status to Browse -> Builds.

**Comment by spadgett** [comment]
@liggitt PR updated when you have time

File: assets/app/views/_triggers.html [hunk 3]
```diff
@@ -1,11 +1,25 @@
...
       <span ng-switch-when="Failed" class="fa fa-times text-danger" aria-hidden="true" style="margin-right: 5px;"></span>
+      <span ng-switch-when="Error" class="fa fa-times text-danger" aria-hidden="true" style="margin-right: 5px;"></span>
+      <span ng-switch-when="Cancelled" class="fa fa-ban text-warning" aria-hidden="true"></span>
       <span ng-switch-default class="fa fa-refresh fa-spin" aria-hidden="true" style="margin-right: 5px;"></span>
     </span>
-    <span>A build of {{build.metadata.labels.buildconfig}} is {{build.status | lowercase}}.</span>
-    <span ng-if="build.status != 'Complete' && build.status != 'Failed' && trigger.imageChangeParams.automatic">  A new deployment will be created automatically once the build completes.</span>
+    <!-- Message -->
+    <span ng-switch="build.status" class="hide-ng-leave">
+      <span ng-switch-when="Complete">A build of {{build.metadata.labels.buildconfig}} completed.</span>
```
Comment (liggitt) [inline assets/app/views/_triggers.html]: factor out `A build of {{build.metadata.labels.buildconfig}}` so we don't have to repeat it. Error can say "... encountered an error."

File: assets/app/views/_triggers.html [hunk 4]
```diff
@@ -1,11 +1,25 @@
...
+      <span ng-switch-when="Failed">A build of {{build.metadata.labels.buildconfig}} failed.</span>
+      <span ng-switch-when="Error">An error prevented a build of {{build.metadata.labels.buildconfig}} from running.</span>
+      <span ng-switch-when="Cancelled">A build of {{build.metadata.labels.buildconfig}} was cancelled.</span>
+      <span ng-switch-default>
+        A build of {{build.metadata.labels.buildconfig}} is {{build.status | lowercase}}.
+      </span>
+    </span>
+    <span ng-if="build.status != 'Complete' && build.status != 'Failed' && build.status != 'Cancelled' && trigger.imageChangeParams.automatic">
```
Comment (liggitt) [inline assets/app/views/_triggers.html]: this can nest inside the `ng-switch-default` case above and drop all the `build.status != ...` checks

**Comment by spadgett** [comment]
@liggitt I noticed a problem where Cancelled builds were not cleared from the Overview page after 5 minutes. I've fixed that now and updated the PR with your suggestions.

File: assets/app/views/_triggers.html [hunk 5]
```diff
@@ -1,11 +1,25 @@
 <div class="builds" ng-repeat="trigger in triggers">
-  <div ng-repeat="build in trigger.builds" ng-if="(build.status != 'Complete' && build.status != 'Failed') || (build.metadata.creationTimestamp | ageLessThan : 5 : 'minutes')" class="build animate-repeat osc-object" kind="Build" resource="build">
+  <div ng-repeat="build in trigger.builds" ng-if="build.status == 'New' || build.status == 'Pending' || build.status == 'Running' || (build.metadata.creationTimestamp | ageLessThan : 5 : 'minutes')" class="build animate-repeat osc-object" kind="Build" resource="build">
```
Comment (liggitt) [inline assets/app/views/_triggers.html]: Probably worth separating the check if the build is running into a filter ("isBuildInProgress", etc) that checks new, pending, running

Comment (liggitt) [inline assets/app/views/_triggers.html]: Fine to do that in a separate PR

**Comment by liggitt** [comment]
[merge]
